llvm/lib/Object/DXContainer.cpp

```diff
@@ -288,6 +289,50 @@ Error DirectX::RootSignature::parse(StringRef Data) {
                             llvm::Twine(FValue));
   Flags = FValue;
 
+  Current = Begin + RootParametersOffset;
+  for (uint32_t It = 0; It < NumParameters; It++) {
+    dxbc::RootParameter NewParam;
+
+    dxbc::RootParameterType PTValue =
+        support::endian::read<dxbc::RootParameterType,
+                              llvm::endianness::little>(Current);
+    if (!dxbc::RootSignatureValidations::isValidParameterType(PTValue))
+      return validationFailed("unsupported parameter type value read: " +
+                              llvm::Twine((uint32_t)PTValue));
+
+    NewParam.ParameterType =
+        support::endian::read<dxbc::RootParameterType,
+                              llvm::endianness::little>(Current);
+    Current += sizeof(dxbc::RootParameterType);
+
+    dxbc::ShaderVisibilityFlag SVValue =
+        support::endian::read<dxbc::ShaderVisibilityFlag,
+                              llvm::endianness::little>(Current);
+
+    if (!dxbc::RootSignatureValidations::isValidShaderVisibility(SVValue))
+      return validationFailed("unsupported shader visility flag value read: " +
+                              llvm::Twine((uint32_t)SVValue));
+
+    NewParam.ShaderVisibility = SVValue;
+    Current += sizeof(dxbc::ShaderVisibilityFlag);
+
+    uint32_t Offset =
+        support::endian::read<uint32_t, llvm::endianness::little>(Current);
+    Current += sizeof(uint32_t);
+
+    switch (NewParam.ParameterType) {
+
+    case dxbc::RootParameterType::Constants32Bit: {
+      if (Error Err = readStruct(Data, Begin + Offset, NewParam.Constants))
+        return Err;
+    } break;
+    case dxbc::RootParameterType::Empty:
+      llvm_unreachable("Invalid value for RootParameterType");
```

Comment:
Maybe add a note here that this is unreachable because it was validated earlier?